src/board.rb

```diff
@@ -0,0 +1,84 @@
+class Board
+
+  DIRECTION_MATRIX = [
+      [1, 0],
+      [-1, 0],
+      [0, 1],
+      [0, -1],
+      [1, 1],
+      [-1, -1],
+      [-1, 1],
+      [1, -1]
+  ]
+
+  # @param [Integer] num_columns
+  # @param [Integer] num_cells
+  def initialize(num_columns, num_cells)
+    @num_columns = num_columns
+    @num_cells = num_cells
+    @board = num_columns.times.map {[]}
+  end
+
+  def to_s
+    out = ['-' * (@num_columns * 2 + 3)]
+    @num_cells.times do |row|
+      line = '| '
+      @num_columns.times do |col|
+        line << '%s ' % (@board[col][@num_cells - row - 1] || ' ')
+      end
+      line << '|'
+      out << line
+    end
+    out << '-' * (@num_columns * 2 + 3)
+    out << '  %s' % @num_columns.times.map {|i| '%d ' % i}.join
+    out.join("\n")
+  end
+
+  # @param [Integer] column_index
+  # @param [String] player_symbol
+  def move(column_index, player_symbol)
+    raise RangeError.new('Column index should be from %d to %d' % [0, @num_columns - 1]) if column_index > @num_columns - 1
+    raise RangeError.new('Column #%d is full' % column_index) unless @board[column_index].length < @num_cells
+    @board[column_index] << player_symbol
+  end
+
+  # @param [Integer] column_index
+  def cancel_move(column_index)
+    @board[column_index].pop
+  end
+
+  def available_moves
+    @num_columns.times.select {|col| @board[col].count < @num_cells}
+  end
+
+
+  # @param [Integer] col
+  # @param [Integer] num
+  def detect_winner(col, num)
```

Comment:
Can you comment on the implementation of this method? I'm interested in learning a bit on how you feel about this part of the implementation. Do you like it? If so what's good about it? If not, what can be improved?

Reply:
This is actually a second implementation. First one did not depend on the last move and was able to find winning combination on a board in any state. But I decided that it's inefficient since we always know that there was no winning combination before the last move. And we only need to look for a winning combination that includes the last move. Hence the solution, it takes the last move (col index, row index) and "walks" over the cells in 8 directions counting symbols as long as they match the initial one. 

So generally speaking I like this approach. 

As for the implementation, I would say it's definitely can be improved. It's not very readable right now. I was thinking on extracting helper lambda into a private method and maybe making the counting structure look more explicit (right now it's ```[[[0],[0]],[[0],[0]],[[0],[0]],[[0],[0]]]``` which is something I would expect people to not understand easily). So if I had more time for this, I would improve this part for sure.